// mexc-futures-sdk/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatErrorForLogging = exports.parseAxiosError = exports.MexcRateLimitError = exports.MexcSignatureError = exports.MexcValidationError = exports.MexcNetworkError = exports.MexcApiError = exports.MexcAuthenticationError = exports.MexcFuturesError = exports.MexcFuturesWebSocket = exports.MexcFuturesClient = void 0;
var client_1 = require("./client");
Object.defineProperty(exports, "MexcFuturesClient", { enumerable: true, get: function () { return client_1.MexcFuturesSDK; } });
// Export WebSocket client
var websocket_1 = require("./websocket");
Object.defineProperty(exports, "MexcFuturesWebSocket", { enumerable: true, get: function () { return websocket_1.MexcFuturesWebSocket; } });
// Export error classes
var errors_1 = require("./utils/errors");
Object.defineProperty(exports, "MexcFuturesError", { enumerable: true, get: function () { return errors_1.MexcFuturesError; } });
Object.defineProperty(exports, "MexcAuthenticationError", { enumerable: true, get: function () { return errors_1.MexcAuthenticationError; } });
Object.defineProperty(exports, "MexcApiError", { enumerable: true, get: function () { return errors_1.MexcApiError; } });
Object.defineProperty(exports, "MexcNetworkError", { enumerable: true, get: function () { return errors_1.MexcNetworkError; } });
Object.defineProperty(exports, "MexcValidationError", { enumerable: true, get: function () { return errors_1.MexcValidationError; } });
Object.defineProperty(exports, "MexcSignatureError", { enumerable: true, get: function () { return errors_1.MexcSignatureError; } });
Object.defineProperty(exports, "MexcRateLimitError", { enumerable: true, get: function () { return errors_1.MexcRateLimitError; } });
Object.defineProperty(exports, "parseAxiosError", { enumerable: true, get: function () { return errors_1.parseAxiosError; } });
Object.defineProperty(exports, "formatErrorForLogging", { enumerable: true, get: function () { return errors_1.formatErrorForLogging; } });
// Export types
__exportStar(require("./types/orders"), exports);
__exportStar(require("./types/account"), exports);
__exportStar(require("./types/market"), exports);
__exportStar(require("./utils/headers"), exports);
